File: 11/sol.py
import numpy as np
from functools import reduce
from pprint import pprint

def part1(grid, steps):
    dim = len(grid)
    total_flashes = 0
    def flash_at(_grid, iy, ix, flashes):
        if iy >= dim or iy < 0 or ix <0 or ix >=dim:
            return _grid, flashes
        val_at = _grid[iy][ix]
        if val_at == False:
            return _grid, flashes
        elif val_at > 9:
            _grid[iy][ix] = False
            flashes += 1
            # recurse
            nextPts = [(iy+y, ix+x) for x in [-1,0,1] for y in [-1,0,1]]
            newGrid = _grid
            for ny, nx in nextPts:
                newGrid, flashes = flash_at(newGrid, ny, nx, flashes)
            return newGrid, flashes
        else:
            _grid[iy][ix] += 1
            if val_at + 1 > 9:
                return flash_at(_grid, iy, ix, flashes)
            return _grid, flashes
            
    flash_counts = []
    # Actual execution
    for _ in range(steps):
        flashes_at_step = 0
        for iy, ix in np.ndindex((dim, dim)):
            grid[iy][ix] += 1
        # Compute flash
        for iy, ix in np.ndindex((dim, dim)):
            if grid[iy][ix] > 9:
                grid,flashes_at_step= flash_at(grid, iy, ix, flashes_at_step)
                 
        for iy, ix in np.ndindex((dim, dim)):
            if grid[iy][ix] == False:
                grid[iy][ix] = 0
        flash_counts.append(flashes_at_step)
        #print("step", s+1)
        #pprint(grid)
    return flash_counts

def part2(grid):
    dim = len(grid)
    total_flashes = 0
    def flash_at(_grid, iy, ix, flashes):
        if iy >= dim or iy < 0 or ix <0 or ix >=dim:
            return _grid, flashes
        val_at = _grid[iy][ix]
        if val_at == False:
            return _grid, flashes
        elif val_at > 9:
            _grid[iy][ix] = False
            flashes += 1
            # recurse
            nextPts = [(iy+y, ix+x) for x in [-1,0,1] for y in [-1,0,1]]
            newGrid = _grid
            for ny, nx in nextPts:
                newGrid, flashes = flash_at(newGrid, ny, nx, flashes)
            return newGrid, flashes
        else:
            _grid[iy][ix] += 1
            if val_at + 1 > 9:
                return flash_at(_grid, iy, ix, flashes)
            return _grid, flashes
            
    # Actual execution
    s = 0
    while True:
        s += 1
        flashes_at_step = 0
        for iy, ix in np.ndindex((dim, dim)):
            grid[iy][ix] += 1
        # Compute flash
        for iy, ix in np.ndindex((dim, dim)):
            if grid[iy][ix] > 9:
                grid,flashes_at_step= flash_at(grid, iy, ix, flashes_at_step)
                 
        for iy, ix in np.ndindex((dim, dim)):
            if grid[iy][ix] == False:
                grid[iy][ix] = 0
        if flashes_at_step == dim*dim:

            return s

if __name__ == "__main__":
    
    file_input = []
    with open("input.txt","r") as f:
        for l in f:
            file_input.append(list(map(int, l.rstrip())))
    print(sum(part1(file_input, 100)))
    print(part2(file_input) + 100)

File: 11/sol.hs
import System.IO()
import Text.Printf
import Data.List.Index
import Data.Array

-- Read solution and pass it on
main :: IO ()

type Grid = Array Int (Array Int (Int, Bool))


computeFlashes :: Grid -> Int -> [(Grid, Int)]
computeFlashes grid dim
  = scanl (\(g, _) _ -> let (g', f') = computeIter g in (g',f')) (grid, 0) (repeat 0)
  where
    flashAt :: Grid -> Int -> (Int, Int) -> (Grid, Int)
    flashAt g flashes pt@(y, x)
      | x >= dim || x < 0 || y >= dim || y < 0 = (g, flashes) -- Checking bounds
      | state                                  = (g, flashes)
      | num > 9                                = adjFlashes
      | num + 1 > 9                            = flashAt incGrid flashes pt
      | otherwise                              = (incGrid, flashes)
      where
        row'         = g ! y
        (num, state) = (row') ! x

        -- Add one to the current point
        incGrid      = g // [(y, row' // [(x, (num + 1, state))])]
        flipGrid     = g // [(y, row' // [(x, (0, True))])]

        --- Recursive steps:
        nextPts      = [(y + ix, x + iy) | ix <- [-1, 0, 1], iy <- [-1, 0, 1]]
        adjFlashes   = foldl (\(g', f') pt' -> flashAt g' f' pt') (flipGrid, flashes + 1) nextPts

    computeIter :: Grid -> (Grid, Int)
    computeIter g
      = (zeroedout, flashes)
      where
        pts = [(y, x) | x <- [0..(dim - 1)], y <- [0..(dim - 1)]]
        (flashed, flashes) = foldl (\(g', f) pt -> flashAt g' f pt ) (g, 0) pts
        zeroedout = flashed // [(y, (flashed ! y) // [(x, (fst ((flashed ! y) ! x), False)) | x <- [0..(dim -1 )] ] )
                               | y <- [0..(dim - 1)]]
    

main = do
    contentLines <- map (\x -> [c:[] | c <- x]) <$> lines <$> readFile "input.txt"

    -- Parsing Here
    let n       = length contentLines
        inputs' = map (map ((\x -> (x, False)) . read)) contentLines :: [[(Int, Bool)]]
        inputs'' = map (listArray (0, n - 1)) inputs'
        input    = listArray (0, n - 1) inputs''
        flashes  = computeFlashes input n
        part1    = sum $ map snd $ take 101 flashes
        part2    = head $ dropWhile (\(_, x) -> x /= n * n) $ indexed (map snd $ flashes)
        
    printf "Part 1: %d\n" part1
    printf "Part 2: %d\n" $ fst part2
